File: 프로그래머스/lv2/42626. 더 맵게/더 맵게.py
import heapq

def solution(scoville, K):
    answer = 0
    heap = []
    
    for s in scoville:
        heapq.heappush(heap, s)
        
    while heap[0] < K:
        answer += 1
        s1 = heapq.heappop(heap)
        s2 = heapq.heappop(heap)
        heapq.heappush(heap, s1 + s2*2)
        
        if len(heap) == 1 and heap[0] < K:
            return -1
        
    return answer

File: 프로그래머스/lv2/42626. 더 맵게/더 맵게.java
import java.util.*;

class Solution {
    public int solution(int[] scoville, int K) {
        int answer = 0;
        Queue<Integer> pq = new PriorityQueue<Integer>();
        
        for(int s : scoville){
            pq.offer(s);
        }
        
        while (pq.peek() < K){
            answer++;
            int s1 = pq.poll();
            int s2 = pq.poll();
            pq.offer(s1 + s2*2);
            
            if(pq.size() == 1 && pq.peek() < K)
                return -1;
        }
        
        return answer;
    }
}